<a href="https://colab.research.google.com/github/ErnestoFranCh/Repaso/blob/main/Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

#Lectura y escritura de datos en Pandas

## Lectura y escritura de datos
- **`pd.read_csv()`**: Lee un archivo CSV y lo convierte en un DataFrame.
- **`pd.read_excel()`**: Lee archivos de Excel y los convierte en DataFrames.
- **`pd.read_sql()`**: Ejecuta consultas SQL y devuelve los resultados en un DataFrame.
- **`DataFrame.to_csv()`**: Escribe un DataFrame en un archivo CSV.
- **`DataFrame.to_excel()`**: Escribe un DataFrame en un archivo de Excel.

Para mas informacion sobre la lectura de archivos con pandas entrar al siguiente link [IO tools (text, CSV, HDF5, …)](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#)
## `pd.read_excel()`


```python
df = pd.read_excel('archivo.xlsx', sheet_name='Hoja1')
```


## `DataFrame.to_excel()`
```python
df.to_excel('salida.xlsx', sheet_name='Resultados', index=False)
```



In [ ]:
EmployeeData = pd.read_excel('/content/drive/MyDrive/Data colab notes/Employee-Management-Sample-Data_clean.xlsx')

# Exploracion de datos
 -  `.head(n)`	Muestra las primeras n filas (por defecto, 5).
 -  `.tail(n)`	Muestra las últimas n filas.
 -	`.info()` Muestra información sobre el DataFrame (tipos de datos, memoria, valores nulos).
 - `.describe()`	Genera estadísticas descriptivas de las columnas numéricas.
 - `.shape`	Devuelve una tupla con el número de filas y columnas.
 - `.columns`	Lista los nombres de las columnas.
 - `.dtypes`	Muestra los tipos de datos de cada columna.
 - `.unique()`	Devuelve los valores únicos de una columna.
 - `.nunique()`	Cuenta los valores únicos en una columna.
 - `.value_counts()` 	Cuenta la frecuencia de cada valor en una columna.



In [ ]:
EmployeeData.head() #primeros 5 registros
EmployeeData.head(10) #primeros 10 registros

Employee ID           Full Name                Department  \
0       S1001          John Smith           Human Resources   
1       S1002            Jane Doe                 Marketing   
2       S1003     Michael Johnson                   Finance   
3       S1004         Emily Brown                     Sales   
4       S1005        David Wilson    Information Technology   
5       S1006         Lisa Taylor                Operations   
6       S1007     Daniel Martinez          Customer Service   
7       S1008      Sarah Anderson  Research and Development   
8       S1009  Christopher Thomas                Production   
9       S1010     Kimberly Garcia         Quality Assurance   

                       Designation  Hire Date  Annual Salary (USD)  
0                       HR Manager 2023-05-15                60000  
1             Marketing Specialist 2023-08-20                50000  
2                Financial Analyst 2023-10-10                55000  
3             Sales Representative 2023-03-25                48000  
4                    IT Specialist 2023-09-12                60000  
5               Operations Manager 2023-06-30                65000  
6  Customer Service Representative 2023-11-05                45000  
7                     R&D Engineer 2023-04-18                58000  
8            Production Supervisor 2023-07-22                52000  
9                       QA Analyst 2023-02-14                56000

In [ ]:
EmployeeData.tail() #Ultimos cinco registros
EmployeeData.tail(10) #ultimos 10 registros

Employee ID          Full Name                Department  \
30       S1031      Michael Smith           Human Resources   
31       S1032    Jessica Johnson                 Marketing   
32       S1033        David Brown                   Finance   
33       S1034       Sarah Garcia                     Sales   
34       S1035  Matthew Hernandez    Information Technology   
35       S1036    Emily Rodriguez                Operations   
36       S1037       Daniel Davis          Customer Service   
37       S1038     Jennifer Smith  Research and Development   
38       S1039    Michael Johnson                Production   
39       S1040   Jessica Martinez         Quality Assurance   

                   Designation  Hire Date  Annual Salary (USD)  
30               HR Generalist 2023-07-17                50000  
31         Marketing Assistant 2023-04-10                42000  
32    Junior Financial Analyst 2023-11-26                48000  
33           Sales Coordinator 2023-06-17                50000  
34        Junior IT Specialist 2023-09-20                48000  
35        Operations Assistant 2023-12-03                42000  
36  Customer Service Associate 2023-04-06                40000  
37               R&D Assistant 2023-07-30                42000  
38        Production Assistant 2023-10-25                40000  
39                QA Assistant 2023-03-20                42000

In [ ]:
EmployeeData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40 entries, 0 to 39
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Employee ID          40 non-null     object        
 1   Full Name            40 non-null     object        
 2   Department           40 non-null     object        
 3   Designation          40 non-null     object        
 4   Hire Date            40 non-null     datetime64[ns]
 5   Annual Salary (USD)  40 non-null     int64         
dtypes: datetime64[ns](1), int64(1), object(4)
memory usage: 2.0+ KB


Interpretacion:

- `<class 'pandas.core.frame.DataFrame>` : Te indica que estás trabajando con un objeto  `DataFrame`.

- `RangeIndex: 40 entries, 0 to 39` : Hay 40 filas, con índices del 0 al 39.

- `Data columns (total 6 columns):` : El `DataFrame` tiene 6 columnas en total.

- Tabla con informacion por columna.

- `dtypes: datetime64[ns](1), int64(1), object(4)` : Resumen de los tipos de datos que tiene el  `DataFrame``

- `memory usage: 2.0+ KB`:	Memoria total estimada que ocupa el `DataFrame`.

    - Se refiere a la cantidad de memoria `RAM ` que el `DataFrame` está
       utilizando en tu entorno de ejecución de Python.



## `DataFrame.describe(percentiles=None, include=None, exclude=None)`

- `include` : Este argumento te permite incluir tipos específicos de columnas en el resumen.

  - `'None'` : Solo columnas numericas.

  - `'all'` : todo tipo de columnas, fechas, numericas,categoricas, etc.

  - `'object'` : Solo columnas de tipo texto o categorico

  - Lista combinada: `include = ['object','number']`

- `percentiles` : Este argumento te deja modificar los percentiles que quieres que aparezcan en el resumen.

- `exclude` : Caso contrario de `include`

     


In [ ]:
EmployeeData.describe()

Hire Date  Annual Salary (USD)
count                   40            40.000000
mean   2023-07-10 13:12:00         55150.000000
min    2023-01-17 00:00:00         40000.000000
25%    2023-04-25 12:00:00         48000.000000
50%    2023-07-11 00:00:00         53500.000000
75%    2023-09-22 12:00:00         62750.000000
max    2023-12-03 00:00:00         75000.000000
std                    NaN         10654.359888

In [ ]:
EmployeeData.describe(include= 'object')

Employee ID        Full Name       Department Designation
count           40               40               40          40
unique          40               37               10          40
top          S1001  Michael Johnson  Human Resources  HR Manager
freq             1                2                4           1

In [ ]:
EmployeeData.describe(include='datetime')

Hire Date
count                   40
mean   2023-07-10 13:12:00
min    2023-01-17 00:00:00
25%    2023-04-25 12:00:00
50%    2023-07-11 00:00:00
75%    2023-09-22 12:00:00
max    2023-12-03 00:00:00

In [ ]:
EmployeeData.describe(include='number',percentiles=[0.1, 0.5, 0.9])

Annual Salary (USD)
count            40.000000
mean          55150.000000
std           10654.359888
min           40000.000000
10%           42000.000000
50%           53500.000000
90%           70200.000000
max           75000.000000

In [ ]:
EmployeeData.shape #tupla con las (filas,col)

print('filas = {} \nColumnas = {}'
                                  .format(EmployeeData.shape[0],EmployeeData.shape[1]))

filas = 40 
Columnas = 6


In [ ]:
lst = list(EmployeeData.columns)
lst

['Employee ID',
 'Full Name',
 'Department',
 'Designation',
 'Hire Date',
 'Annual Salary (USD)']

In [ ]:
EmployeeData.dtypes

Employee ID                    object
Full Name                      object
Department                     object
Designation                    object
Hire Date              datetime64[ns]
Annual Salary (USD)             int64
dtype: object

In [ ]:
unicos = list(EmployeeData.Department.unique()) #Devuelve array de numpy
print(len(unicos))
unicos

10


['Human Resources',
 'Marketing',
 'Finance',
 'Sales',
 'Information Technology',
 'Operations',
 'Customer Service',
 'Research and Development',
 'Production',
 'Quality Assurance']

In [ ]:
EmployeeData.Department.nunique()

10

In [ ]:
EmployeeData.Department.value_counts()

Department
Human Resources             4
Marketing                   4
Finance                     4
Sales                       4
Information Technology      4
Operations                  4
Customer Service            4
Research and Development    4
Production                  4
Quality Assurance           4
Name: count, dtype: int64

# Creación de estructuras de datos
- **`pd.Series()`**: Crea una Serie, que es una estructura unidimensional similar a una lista.
- **`pd.DataFrame()`**: Crea un DataFrame, que es una estructura bidimensional similar a una tabla.



## `pd.Series()`



```
pd.Series(data=None, index=None, dtype=None, name=None, copy=False, fastpath=False)
```

- `data` :	Los datos que contendrá la Serie (lista, array, dict, escalar, etc.).

- `index` :	Etiquetas del índice. Si no se proporciona, se genera automáticamente (0, 1, 2...).

- `dtype`: Tipo de datos (int, float, str, etc.). Opcional.

- `name`:	Nombre de la Serie.

- `copy` :	Si se copia o no los datos originales.


In [ ]:
datos = [10, 20, 30, 40]
serie = pd.Series(data=datos)
print(serie)

serie_in = pd.Series(datos, index = ['a','b','c','d'],name= 'serie_indices') #name da el nombre a la serie como si fuera una columna
print(serie_in)

0    10
1    20
2    30
3    40
dtype: int64
a    10
b    20
c    30
d    40
Name: serie_indices, dtype: int64


## `pd.DataFrame()`



```python
pd.DataFrame(data=None, index=None, columns=None, dtype=None, copy=None)

```

- `columns` : 	Nombres de las columnas (opcional). `columns = ['name1' ,'name2',..]`


In [ ]:
 data = {
    'Nombre': ['Ana', 'Luis', 'Carlos'],
    'Edad': [23, 34, 45],
    'Ciudad': ['CDMX', 'GDL', 'MTY']
}

df = pd.DataFrame(data)
df

Nombre  Edad Ciudad
0     Ana    23   CDMX
1    Luis    34    GDL
2  Carlos    45    MTY

In [ ]:
data = [[1, 2], [3, 4]]
df = pd.DataFrame(data, index=['a', 'b'], columns=['x', 'y'])
print(df)

   x  y
a  1  2
b  3  4


In [ ]:
data = [
    {'Nombre': 'Ana', 'Edad': 23},
    {'Nombre': 'Luis', 'Edad': 34}
]
df = pd.DataFrame(data)
print(df)

  Nombre  Edad
0    Ana    23
1   Luis    34


In [ ]:
s = pd.Series([100, 200, 300], name='Ventas')
df = pd.DataFrame(s)
df

Ventas
0     100
1     200
2     300

# Selección y filtrado de datos
- **`DataFrame.loc[]`**: Selecciona datos por etiquetas o una condición booleana.
- **`DataFrame.iloc[]`**: Selecciona datos por índices de posición.
- **`DataFrame.at[]`**: Accede a un solo valor utilizando una etiqueta.
- **`DataFrame.iat[]`**: Accede a un solo valor utilizando índices de posición.

## **`DataFrame.loc[]`**



```python
df.loc[filas, columnas]
```



In [ ]:
EmployeeData.head(1)

Employee ID   Full Name       Department Designation  Hire Date  \
0       S1001  John Smith  Human Resources  HR Manager 2023-05-15   

   Annual Salary (USD)  
0                60000

### Ejemplos de vistas:

In [ ]:
EmployeeData.loc[:,'Full Name'] # Solo una columna
EmployeeData.loc[:,['Employee ID','Department']] #Varias columnas

#.loc[:,'col'] el simbolo ':' indica todas las filas

In [ ]:
EmployeeData.loc[0:2,'Full Name']

0         John Smith
1           Jane Doe
2    Michael Johnson
Name: Full Name, dtype: object

In [ ]:
EmployeeData.loc[0] #La primer fila

Employee ID                          S1001
Full Name                       John Smith
Department                 Human Resources
Designation                     HR Manager
Hire Date              2023-05-15 00:00:00
Annual Salary (USD)                  60000
Name: 0, dtype: object

In [ ]:
type(EmployeeData.loc[0,'Annual Salary (USD)'])  #Devuelve un unico valor

numpy.int64

In [ ]:
int(EmployeeData.loc[0,'Annual Salary (USD)'])

60000

In [ ]:
EmployeeData.loc[[0,1,2],['Annual Salary (USD)','Employee ID']] #Seleccionar multiples columnas y filas

Annual Salary (USD) Employee ID
0                60000       S1001
1                50000       S1002
2                55000       S1003

### Condicion Booleana.



```python
df.loc[condición, columnas_opcionales]
```



In [ ]:
#Filtra todas las filas y columnas donde Annual Salary (USD) sea mayor a 55000
EmployeeData.loc[EmployeeData['Annual Salary (USD)']> 55000]

Employee ID              Full Name                Department  \
0        S1001             John Smith           Human Resources   
4        S1005           David Wilson    Information Technology   
5        S1006            Lisa Taylor                Operations   
7        S1008         Sarah Anderson  Research and Development   
9        S1010        Kimberly Garcia         Quality Assurance   
11       S1012          Melissa Lopez                 Marketing   
12       S1013          Richard Perez                   Finance   
13       S1014       Jessica Gonzalez                     Sales   
14       S1015         Matthew Wilson    Information Technology   
16       S1017          James Johnson          Customer Service   
17       S1018            Laura Brown  Research and Development   
18       S1019           Daniel Smith                Production   
19       S1020         Jennifer Davis         Quality Assurance   
22       S1023  Christopher Rodriguez                   Finance   
24       S1025           David Wilson    Information Technology   
28       S1029          Daniel Wilson                Production   
29       S1030      Jennifer Martinez         Quality Assurance   

                     Designation  Hire Date  Annual Salary (USD)  
0                     HR Manager 2023-05-15                60000  
4                  IT Specialist 2023-09-12                60000  
5             Operations Manager 2023-06-30                65000  
7                   R&D Engineer 2023-04-18                58000  
9                     QA Analyst 2023-02-14                56000  
11             Marketing Manager 2023-06-10                65000  
12               Finance Manager 2023-11-18                70000  
13                 Sales Manager 2023-05-03                72000  
14                    IT Manager 2023-08-08                75000  
16      Customer Service Manager 2023-03-12                68000  
17                   R&D Manager 2023-06-25                70000  
18            Production Manager 2023-09-30                72000  
19                    QA Manager 2023-01-17                75000  
22      Senior Financial Analyst 2023-11-12                60000  
24          Senior IT Specialist 2023-08-30                70000  
28  Senior Production Supervisor 2023-09-08                60000  
29             Senior QA Analyst 2023-02-22                62000

In [ ]:
EmployeeData.loc[EmployeeData['Annual Salary (USD)']>55000, ['Full Name','Annual Salary (USD)']]
#Filtra pero solo muestra las columnas Full Name y Annual Salary (USD)

Full Name  Annual Salary (USD)
0              John Smith                60000
4            David Wilson                60000
5             Lisa Taylor                65000
7          Sarah Anderson                58000
9         Kimberly Garcia                56000
11          Melissa Lopez                65000
12          Richard Perez                70000
13       Jessica Gonzalez                72000
14         Matthew Wilson                75000
16          James Johnson                68000
17            Laura Brown                70000
18           Daniel Smith                72000
19         Jennifer Davis                75000
22  Christopher Rodriguez                60000
24           David Wilson                70000
28          Daniel Wilson                60000
29      Jennifer Martinez                62000

In [ ]:
EmployeeData.loc[EmployeeData['Full Name'].str.startswith('A'), ['Full Name', 'Annual Salary (USD)']]
## Muestra solo aquellas personas cuyo nombre empieza con A

Full Name  Annual Salary (USD)
15  Amanda Martinez                48000
21    Amy Hernandez                48000

### Modificar valores

In [ ]:
EmployeeData.loc[15,'Annual Salary (USD)']= 55000

EmployeeData.loc[EmployeeData['Full Name'].str.startswith('A'), ['Full Name', 'Annual Salary (USD)']]

Full Name  Annual Salary (USD)
15  Amanda Martinez                55000
21    Amy Hernandez                48000

## **`DataFrame.iloc[]`**



```python
df.iloc[fila, columna]

```



In [ ]:
EmployeeData.iloc[0] #Primera fila

Employee ID                          S1001
Full Name                       John Smith
Department                 Human Resources
Designation                     HR Manager
Hire Date              2023-05-15 00:00:00
Annual Salary (USD)                  60000
Name: 0, dtype: object

In [ ]:
EmployeeData.iloc[:,0] #Primer columna

In [ ]:
EmployeeData.iloc[0:3] #Primeras tres filas

Employee ID        Full Name       Department           Designation  \
0       S1001       John Smith  Human Resources            HR Manager   
1       S1002         Jane Doe        Marketing  Marketing Specialist   
2       S1003  Michael Johnson          Finance     Financial Analyst   

   Hire Date  Annual Salary (USD)  
0 2023-05-15                60000  
1 2023-08-20                50000  
2 2023-10-10                55000

In [ ]:
EmployeeData.iloc[0:3,0:3] # primeras dos filas y primeras dos columnas
EmployeeData.iloc[2:4,4:6] #de la fila 2 a la 3 y columnas 4 a la 5

Hire Date  Annual Salary (USD)
2 2023-10-10                55000
3 2023-03-25                48000

### Modificar datos.

In [ ]:
EmployeeData.iloc[15,5] = 48000

In [ ]:
EmployeeData.iloc[15:17,[1,5]]

Full Name  Annual Salary (USD)
15  Amanda Martinez                48000
16    James Johnson                68000

| Característica        | `.loc[]`                     | `.iloc[]`                          |
|-----------------------|------------------------------|------------------------------------|
| Usa                  | Etiquetas (nombres)          | Índices numéricos (posiciones)     |
| Tipo de selección     | Inclusiva (ej: `0:2` incluye 2) | Exclusiva en slices (ej: `0:2` no incluye 2) |
| Tolerancia            | Error si no existe la etiqueta | Error si índice fuera de rango     |


##  **`DataFrame.at[]`**

Diseñado para acceder y modificar un solo valor de un DataFrame, de manera rápida y eficiente.

`at[]` es más rápido que `.loc[]` y `.iloc[]` cuando se trata de acceder o modificar un solo valor.

```python
#sintaxis
df.at[fila, columna]
```



In [ ]:
EmployeeData.at[15,'Annual Salary (USD)'] = 55000 # Es por nombre de columna

EmployeeData.at[15,'Annual Salary (USD)']

15    55000
16    68000
17    70000
Name: Annual Salary (USD), dtype: int64

In [ ]:
EmployeeData.at[15,'Annual Salary (USD)'] = 48000

EmployeeData.iloc[14:17,5]

14    75000
15    48000
16    68000
Name: Annual Salary (USD), dtype: int64

Full Name  Annual Salary (USD)
15  Amanda Martinez                55000
16    James Johnson                68000

### **`DataFrame.iat[]`**

Útil cuando necesitas acceder o modificar un único valor en un `DataFrame` usando índices de posición enteros (como si fuera una matriz de `Numpy`).



```python
#sintaxis
df.iat[fila, columna]

```

In [ ]:
EmployeeData.iat[1,5]

np.int64(50000)

In [ ]:
EmployeeData.iat[1,5] = 48000
EmployeeData.iat[1,5]

np.int64(48000)

| Característica        | `.at[]`                     | `.iat[]`                          |
|-----------------------|-----------------------------|-----------------------------------|
| Acceso por            | Etiqueta (nombre)           | Índice (posición entera)         |
| Uso                   | Muy legible                 | Más rápido y directo             |
| Soporta               | Solo un valor               | Solo un valor                    |


# Manipulación de datos
- **`DataFrame.drop()`**: Elimina filas o columnas del DataFrame.
- **`DataFrame.rename()`**: Renombra etiquetas de filas o columnas.
- **`DataFrame.sort_values()`**: Ordena el DataFrame por los valores de una o más columnas.
- **`DataFrame.sort_index()`**: Ordena el DataFrame por sus índices.
- **`DataFrame.merge()`**: Combina dos DataFrames en función de una clave común.
- **`DataFrame.groupby()`**: Agrupa datos para operaciones de agregación.

## DataFrame.drop()



```python
DataFrame.drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise')

```

- `labels` :	El nombre o lista de nombres (de filas o columnas) a eliminar.

- `axis`:	`0` para filas (default), `1` para columnas.
    - `axis=0` → fila por fila (↓)

    - `axis=1` → columna por columna (→)

- `index`:	Alias para labels cuando axis=0.

- `columns`	Alias para labels cuando axis=1.

- `inplace`:	`True` para modificar el DataFrame original.

- `errors` :	`raise` lanza error si no se encuentra el label, `ignore` no lanza error.



In [ ]:
Prueba = EmployeeData.head(2).copy()

# Elimina la fila con índice 0 (no modifica df original), por lo que, hay que guardar el cambio
Prueba.drop(0)

In [ ]:
eliminacion = Prueba.drop(0)  # Elimina la fila con índice 0 y el resultado se guarda

eliminacion

Employee ID Full Name Department           Designation  Hire Date  \
1       S1002  Jane Doe  Marketing  Marketing Specialist 2023-08-20   

   Annual Salary (USD)   5  
1                48000 NaN

In [ ]:
#Opcion con inplace

Prueba.drop(0, inplace=True)
Prueba

Employee ID Full Name Department           Designation  Hire Date  \
1       S1002  Jane Doe  Marketing  Marketing Specialist 2023-08-20   

   Annual Salary (USD)   5  
1                48000 NaN

In [ ]:
Prueba2 = EmployeeData.head(10)
Prueba2

Employee ID           Full Name                Department  \
0       S1001          John Smith           Human Resources   
1       S1002            Jane Doe                 Marketing   
2       S1003     Michael Johnson                   Finance   
3       S1004         Emily Brown                     Sales   
4       S1005        David Wilson    Information Technology   
5       S1006         Lisa Taylor                Operations   
6       S1007     Daniel Martinez          Customer Service   
7       S1008      Sarah Anderson  Research and Development   
8       S1009  Christopher Thomas                Production   
9       S1010     Kimberly Garcia         Quality Assurance   

                       Designation  Hire Date  Annual Salary (USD)   5  
0                       HR Manager 2023-05-15                60000 NaN  
1             Marketing Specialist 2023-08-20                48000 NaN  
2                Financial Analyst 2023-10-10                55000 NaN  
3             Sales Representative 2023-03-25                48000 NaN  
4                    IT Specialist 2023-09-12                60000 NaN  
5               Operations Manager 2023-06-30                65000 NaN  
6  Customer Service Representative 2023-11-05                45000 NaN  
7                     R&D Engineer 2023-04-18                58000 NaN  
8            Production Supervisor 2023-07-22                52000 NaN  
9                       QA Analyst 2023-02-14                56000 NaN

In [ ]:
#Eliminar varias filas

Prueba2.drop([1,3],axis=0, inplace=True)
Prueba2

<ipython-input-85-dd81ab684856>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba2.drop([1,3], inplace=True)


Employee ID           Full Name                Department  \
0       S1001          John Smith           Human Resources   
2       S1003     Michael Johnson                   Finance   
4       S1005        David Wilson    Information Technology   
5       S1006         Lisa Taylor                Operations   
6       S1007     Daniel Martinez          Customer Service   
7       S1008      Sarah Anderson  Research and Development   
8       S1009  Christopher Thomas                Production   
9       S1010     Kimberly Garcia         Quality Assurance   

                       Designation  Hire Date  Annual Salary (USD)   5  
0                       HR Manager 2023-05-15                60000 NaN  
2                Financial Analyst 2023-10-10                55000 NaN  
4                    IT Specialist 2023-09-12                60000 NaN  
5               Operations Manager 2023-06-30                65000 NaN  
6  Customer Service Representative 2023-11-05                45000 NaN  
7                     R&D Engineer 2023-04-18                58000 NaN  
8            Production Supervisor 2023-07-22                52000 NaN  
9                       QA Analyst 2023-02-14                56000 NaN

In [ ]:
#Eliminar columna
Prueba2.drop('Hire Date',axis=1, inplace=True)
Prueba2

<ipython-input-87-58f702961aad>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba2.drop('Hire Date',axis=1, inplace=True)


Employee ID           Full Name                Department  \
0       S1001          John Smith           Human Resources   
2       S1003     Michael Johnson                   Finance   
4       S1005        David Wilson    Information Technology   
5       S1006         Lisa Taylor                Operations   
6       S1007     Daniel Martinez          Customer Service   
7       S1008      Sarah Anderson  Research and Development   
8       S1009  Christopher Thomas                Production   
9       S1010     Kimberly Garcia         Quality Assurance   

                       Designation  Annual Salary (USD)   5  
0                       HR Manager                60000 NaN  
2                Financial Analyst                55000 NaN  
4                    IT Specialist                60000 NaN  
5               Operations Manager                65000 NaN  
6  Customer Service Representative                45000 NaN  
7                     R&D Engineer                58000 NaN  
8            Production Supervisor                52000 NaN  
9                       QA Analyst                56000 NaN

In [ ]:
#Eliminar varias columnas
Prueba2.drop([5,'Employee ID'],axis=1,inplace=True)
Prueba2

<ipython-input-89-e42dc32e0eb2>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba2.drop([5,'Employee ID'],axis=1,inplace=True)


Full Name                Department  \
0          John Smith           Human Resources   
2     Michael Johnson                   Finance   
4        David Wilson    Information Technology   
5         Lisa Taylor                Operations   
6     Daniel Martinez          Customer Service   
7      Sarah Anderson  Research and Development   
8  Christopher Thomas                Production   
9     Kimberly Garcia         Quality Assurance   

                       Designation  Annual Salary (USD)  
0                       HR Manager                60000  
2                Financial Analyst                55000  
4                    IT Specialist                60000  
5               Operations Manager                65000  
6  Customer Service Representative                45000  
7                     R&D Engineer                58000  
8            Production Supervisor                52000  
9                       QA Analyst                56000

In [ ]:
#Evitar error si no exite la columna
Prueba2.drop('Employee ID',axis=1,inplace=True,errors='ignore')

<ipython-input-90-ab7505952d99>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba2.drop('Employee ID',axis=1,inplace=True,errors='ignore')


In [ ]:
#Elimina toda la columna y fila
Prueba2.drop(index=0, columns='Full Name', inplace=True)
Prueba2

<ipython-input-92-e44de1fe7b04>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba2.drop(index=0, columns='Full Name', inplace=True)


Department                      Designation  \
2                   Finance                Financial Analyst   
4    Information Technology                    IT Specialist   
5                Operations               Operations Manager   
6          Customer Service  Customer Service Representative   
7  Research and Development                     R&D Engineer   
8                Production            Production Supervisor   
9         Quality Assurance                       QA Analyst   

   Annual Salary (USD)  
2                55000  
4                60000  
5                65000  
6                45000  
7                58000  
8                52000  
9                56000

## **`DataFrame.rename()`**



```python
DataFrame.rename(mapper=None, *, index=None, columns=None, axis=None, inplace=False)
```
- `index`:	Diccionario o función para cambiar los índices (filas).

- `columns`:	Diccionario o función para cambiar nombres de columnas.

- `inplace`:	Si `True`, modifica el `DataFrame` original. Si `False` (por defecto), devuelve una copia modificada.

- `mapper`: es un argumento genérico que se puede usar cuando defines `axis`.



```python
#Ejemplos:

df.rename(mapper={'Nombre': 'Full Name'}, axis=1) #se define axis

df.rename(columns={'Nombre': 'Full Name'}) #alternativa
```

- `*` : En la firma de una función en Python, el `*` marca el inicio de los argumentos `keyword-only`. Es decir: Todos los argumentos después del `*` deben ser pasados con nombre.



```python
df.rename(columns={'Nombre': 'Full Name'})

df.rename({'Nombre': 'Full Name'})  
# ❌ después del * no puedes pasarlo sin nombre de variable

```

In [ ]:
Prueba3 = EmployeeData.head(5)
Prueba3

Employee ID        Full Name              Department           Designation  \
0       S1001       John Smith         Human Resources            HR Manager   
1       S1002         Jane Doe               Marketing  Marketing Specialist   
2       S1003  Michael Johnson                 Finance     Financial Analyst   
3       S1004      Emily Brown                   Sales  Sales Representative   
4       S1005     David Wilson  Information Technology         IT Specialist   

   Hire Date  Annual Salary (USD)   5  
0 2023-05-15                60000 NaN  
1 2023-08-20                48000 NaN  
2 2023-10-10                55000 NaN  
3 2023-03-25                48000 NaN  
4 2023-09-12                60000 NaN

In [ ]:
#Prueba3.rename(mapper={'Employee ID': 'ID'}, axis=1, inplace=True)
Prueba3.rename(columns={'Employee ID': 'ID'}, inplace=True)

#son equivalentes
Prueba3

<ipython-input-105-f0d223dab4b8>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba3.rename(columns={'Employee ID': 'ID'}, inplace=True)


ID        Full Name              Department           Designation  \
0  S1001       John Smith         Human Resources            HR Manager   
1  S1002         Jane Doe               Marketing  Marketing Specialist   
2  S1003  Michael Johnson                 Finance     Financial Analyst   
3  S1004      Emily Brown                   Sales  Sales Representative   
4  S1005     David Wilson  Information Technology         IT Specialist   

   Hire Date  Annual Salary (USD)   5  
0 2023-05-15                60000 NaN  
1 2023-08-20                48000 NaN  
2 2023-10-10                55000 NaN  
3 2023-03-25                48000 NaN  
4 2023-09-12                60000 NaN

In [ ]:
Prueba3.rename(mapper = {0: 'primero',1: 'segundo',2: 'tercero'}, inplace=True)
Prueba3.rename(index={3: 'cuarto',4: 'quinto'}, inplace=True)
Prueba3

<ipython-input-110-60f3ca09ccd4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba3.rename(mapper = {0: 'primero',1: 'segundo',2: 'tercero'}, inplace=True)
<ipython-input-110-60f3ca09ccd4>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba3.rename(index={3: 'cuarto',4: 'quinto'}, inplace=True)


ID        Full Name              Department           Designation  \
primero  S1001       John Smith         Human Resources            HR Manager   
segundo  S1002         Jane Doe               Marketing  Marketing Specialist   
tercero  S1003  Michael Johnson                 Finance     Financial Analyst   
cuarto   S1004      Emily Brown                   Sales  Sales Representative   
quinto   S1005     David Wilson  Information Technology         IT Specialist   

         Hire Date  Annual Salary (USD)   5  
primero 2023-05-15                60000 NaN  
segundo 2023-08-20                48000 NaN  
tercero 2023-10-10                55000 NaN  
cuarto  2023-03-25                48000 NaN  
quinto  2023-09-12                60000 NaN

In [ ]:
Prueba3.rename(index={'primero':0}, columns={'Full Name':'Name'}, inplace=True)
Prueba3

<ipython-input-111-e31a75ca2869>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba3.rename(index={'primero':0}, columns={'Full Name':'Name'}, inplace=True)


ID             Name              Department           Designation  \
0        S1001       John Smith         Human Resources            HR Manager   
segundo  S1002         Jane Doe               Marketing  Marketing Specialist   
tercero  S1003  Michael Johnson                 Finance     Financial Analyst   
cuarto   S1004      Emily Brown                   Sales  Sales Representative   
quinto   S1005     David Wilson  Information Technology         IT Specialist   

         Hire Date  Annual Salary (USD)   5  
0       2023-05-15                60000 NaN  
segundo 2023-08-20                48000 NaN  
tercero 2023-10-10                55000 NaN  
cuarto  2023-03-25                48000 NaN  
quinto  2023-09-12                60000 NaN

## **`DataFrame.sort_values()`**



```python
#sintax
DataFrame.sort_values(
    by,
    axis=0,
    ascending=True,
    inplace=False,
    kind='quicksort',
    na_position='last',
    ignore_index=False
)
```
- `by`:	Nombre de la columna (o lista de columnas) por la que quieres ordenar.

- `ascending`:	`True` (por defecto): orden ascendente; `False`: descendente. Acepta lista si usas varias columnas.

- `inplace`:	`False` por defecto. Si es `True`, modifica el DataFrame original.

- `na_position`:	`'last'` (por defecto): los `NaN` van al final. Usa `'first'` si los quieres al inicio.

- `ignore_index`:	Si es `True`, reinicia los índices después de ordenar.

- `kind`:	Algoritmo de ordenamiento: `'quicksort'`, `'mergesort'`, `'heapsort'`.

In [ ]:
Prueba4 = EmployeeData.head(5)
Prueba4.iat[2,6] = 50
Prueba4

Employee ID        Full Name              Department           Designation  \
0       S1001       John Smith         Human Resources            HR Manager   
1       S1002         Jane Doe               Marketing  Marketing Specialist   
2       S1003  Michael Johnson                 Finance     Financial Analyst   
3       S1004      Emily Brown                   Sales  Sales Representative   
4       S1005     David Wilson  Information Technology         IT Specialist   

   Hire Date  Annual Salary (USD)     5  
0 2023-05-15                60000   NaN  
1 2023-08-20                48000   NaN  
2 2023-10-10                55000  50.0  
3 2023-03-25                48000   NaN  
4 2023-09-12                60000   NaN

In [ ]:
Prueba4.sort_values(by='Full Name') #Ascendente por defecto

Employee ID        Full Name              Department           Designation  \
4       S1005     David Wilson  Information Technology         IT Specialist   
3       S1004      Emily Brown                   Sales  Sales Representative   
1       S1002         Jane Doe               Marketing  Marketing Specialist   
0       S1001       John Smith         Human Resources            HR Manager   
2       S1003  Michael Johnson                 Finance     Financial Analyst   

   Hire Date  Annual Salary (USD)     5  
4 2023-09-12                60000   NaN  
3 2023-03-25                48000   NaN  
1 2023-08-20                48000   NaN  
0 2023-05-15                60000   NaN  
2 2023-10-10                55000  50.0

In [ ]:
Prueba4.sort_values(by='Full Name',ascending=False) #Desc

<ipython-input-118-49937c2e4239>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba4.sort_values(by='Full Name',ascending=False,inplace=True) #Desc


Employee ID        Full Name              Department           Designation  \
2       S1003  Michael Johnson                 Finance     Financial Analyst   
0       S1001       John Smith         Human Resources            HR Manager   
1       S1002         Jane Doe               Marketing  Marketing Specialist   
3       S1004      Emily Brown                   Sales  Sales Representative   
4       S1005     David Wilson  Information Technology         IT Specialist   

   Hire Date  Annual Salary (USD)     5  
2 2023-10-10                55000  50.0  
0 2023-05-15                60000   NaN  
1 2023-08-20                48000   NaN  
3 2023-03-25                48000   NaN  
4 2023-09-12                60000   NaN

In [ ]:
Prueba4.sort_values(by= 5,na_position='last' )

<ipython-input-120-c9d194d3fc2d>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba4.sort_values(by= 5,na_position='last', inplace=True)


Employee ID        Full Name              Department           Designation  \
2       S1003  Michael Johnson                 Finance     Financial Analyst   
0       S1001       John Smith         Human Resources            HR Manager   
1       S1002         Jane Doe               Marketing  Marketing Specialist   
3       S1004      Emily Brown                   Sales  Sales Representative   
4       S1005     David Wilson  Information Technology         IT Specialist   

   Hire Date  Annual Salary (USD)     5  
2 2023-10-10                55000  50.0  
0 2023-05-15                60000   NaN  
1 2023-08-20                48000   NaN  
3 2023-03-25                48000   NaN  
4 2023-09-12                60000   NaN

In [ ]:
#Es valido
# Prueba4.sort_values(by=['Employee ID','Annual Salary (USD)'], ascending=[False, False])

Prueba4.sort_values(by='Annual Salary (USD)', ignore_index=True)#Reinicia el indice

Employee ID        Full Name              Department           Designation  \
0       S1004      Emily Brown                   Sales  Sales Representative   
1       S1002         Jane Doe               Marketing  Marketing Specialist   
2       S1003  Michael Johnson                 Finance     Financial Analyst   
3       S1005     David Wilson  Information Technology         IT Specialist   
4       S1001       John Smith         Human Resources            HR Manager   

   Hire Date  Annual Salary (USD)     5  
0 2023-03-25                48000   NaN  
1 2023-08-20                48000   NaN  
2 2023-10-10                55000  50.0  
3 2023-09-12                60000   NaN  
4 2023-05-15                60000   NaN

##`DataFrame.sort_index()`

Ordena un DataFrame por su índice, ya sea filas o columnas.



```python
#sintax
DataFrame.sort_index(
    axis=0,
    level=None,
    ascending=True,
    inplace=False,
    kind='quicksort',
    na_position='last',
    sort_remaining=True,
    ignore_index=False,
    key=None
)
```
- `axis`:	`0` (por defecto): ordena filas por índice; `1`: ordena columnas por sus nombres.

- `ascending`	`True` (por defecto): orden ascendente; `False`: descendente.

- `inplace`	`False` por defecto. Si es `True`, modifica el `DataFrame` original.

- `level`	Si el índice es jerárquico (`MultiIndex`), permite ordenar por un nivel específico.

- `ignore_index`	Si es `True`, reinicia los índices después de ordenar.

- `key`	Función para transformar los valores del índice antes de ordenarlos (útil para personalizar el orden).

In [ ]:
Prueba4.rename(columns={5:'5'},inplace=True)

<ipython-input-131-25e2032413f3>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Prueba4.rename(columns={5:'5'},inplace=True)


In [ ]:
Prueba4.sort_index(ascending=False) #por sus indices

Employee ID        Full Name              Department           Designation  \
4       S1005     David Wilson  Information Technology         IT Specialist   
3       S1004      Emily Brown                   Sales  Sales Representative   
2       S1003  Michael Johnson                 Finance     Financial Analyst   
1       S1002         Jane Doe               Marketing  Marketing Specialist   
0       S1001       John Smith         Human Resources            HR Manager   

   Hire Date  Annual Salary (USD)     5  
4 2023-09-12                60000   NaN  
3 2023-03-25                48000   NaN  
2 2023-10-10                55000  50.0  
1 2023-08-20                48000   NaN  
0 2023-05-15                60000   NaN

In [ ]:
Prueba4.sort_index(axis=1) #por sus columnas

5  Annual Salary (USD)              Department           Designation  \
4   NaN                60000  Information Technology         IT Specialist   
3   NaN                48000                   Sales  Sales Representative   
1   NaN                48000               Marketing  Marketing Specialist   
0   NaN                60000         Human Resources            HR Manager   
2  50.0                55000                 Finance     Financial Analyst   

  Employee ID        Full Name  Hire Date  
4       S1005     David Wilson 2023-09-12  
3       S1004      Emily Brown 2023-03-25  
1       S1002         Jane Doe 2023-08-20  
0       S1001       John Smith 2023-05-15  
2       S1003  Michael Johnson 2023-10-10

| `sort_values()`                                  | `sort_index()`                                      |
|--------------------------------------------------|-----------------------------------------------------|
| Ordena por valores de una o más columnas.        | Ordena por índice (filas o columnas).              |
| Más común para análisis.                         | Útil para organización o estructuras ordenadas.    |


## `DataFrame.merge()`



```python
#sintaxis
pd.merge(left, right, how='inner', on=None, left_on=None, right_on=None,
         left_index=False, right_index=False, suffixes=('_x', '_y'))

```
- `left`, `right`:	Los `DataFrames` a combinar.
    - El `DataFrame` que pongas en `left` es la base en las uniones con `pd.mere()`.

- `how`:	Tipo de unión: `'inner'`, `'outer'`, `'left'`, `'right'`.
    - `inner`: Solo filas con claves coincidentes en ambos `DataFrames`.
    - `'left'` : Todas las filas del DataFrame izquierdo, y lo que coincida del derecho.
    - `'right'`: Todas las filas del derecho, y lo que coincida del izquierdo.
    - `'outer'`: Todas las filas de ambos DataFrames, uniendo donde se pueda.

- `on` :	Columna(s) en común (mismo nombre en ambos).

- `left_on`, `right_on`	Si las columnas clave tienen nombres diferentes.

- `left_index`, `right_index`	Usa los índices como clave de unión.

- `suffixes` :	Sufijos para columnas duplicadas (`_x`, `_y` por defecto)


In [ ]:
df1 = pd.DataFrame({
    'ID': [1, 2, 3],
    'Nombre': ['Ana', 'Luis', 'María']
})

df2 = pd.DataFrame({
    'ID': [2, 3, 4],
    'Edad': [23, 35, 40]
})

In [ ]:
#how = left
pd.merge(df1,df2,how ='left',on='ID')
#Todas las de la izquierda y lo que coincidad del derecho

ID Nombre  Edad
0   1    Ana   NaN
1   2   Luis  23.0
2   3  María  35.0

In [ ]:
#how = inner

pd.merge(df1,df2,how='inner',on='ID') #Solo filas coincidentes

ID Nombre  Edad
0   2   Luis    23
1   3  María    35

In [ ]:
#how = right
pd.merge(df1,df2,how ='right',on='ID')

ID Nombre  Edad
0   2   Luis    23
1   3  María    35
2   4    NaN    40

In [ ]:
#how = outer
pd.merge(df1,df2,how='outer',on='ID')

ID Nombre  Edad
0   1    Ana   NaN
1   2   Luis  23.0
2   3  María  35.0
3   4    NaN  40.0

In [ ]:
df3 = pd.DataFrame({
    'ID_name': [1, 2, 3],
    'Nombre': ['Ana', 'Luis', 'María']
})

df4 = pd.DataFrame({
    'ID': [2, 3, 4],
    'Edad': [23, 35, 40]
})

In [ ]:
pd.merge(df3,df4,how='left',left_on = 'ID_name',right_on='ID')

ID_name Nombre   ID  Edad
0        1    Ana  NaN   NaN
1        2   Luis  2.0  23.0
2        3  María  3.0  35.0

In [ ]:
df5 = pd.DataFrame({
    'ID': [101, 102, 103],
    'Nombre': ['Ana', 'Luis', 'María']
})

df6 = pd.DataFrame({
    'Edad': [25, 30],
    'Ciudad': ['CDMX', 'Guadalajara']
}, index=[101, 103])

In [ ]:
pd.merge(df5,df6,how='left',left_on='ID',right_index=True)

ID Nombre  Edad       Ciudad
0  101    Ana  25.0         CDMX
1  102   Luis   NaN          NaN
2  103  María  30.0  Guadalajara

## `DataFrame.groupby()`

Agrupa las filas de un DataFrame según los valores de una o más columnas. Luego puedes aplicar funciones como `.sum()`, `.mean()`, `.count()`, etc., a cada grupo.



```python
df.groupby('columna').función() # 1 columna

df.groupby(['col1', 'col2',..]).función() #multiples col


```

In [ ]:
EmployeeData.head(2)

Employee ID   Full Name       Department           Designation  Hire Date  \
0       S1001  John Smith  Human Resources            HR Manager 2023-05-15   
1       S1002    Jane Doe        Marketing  Marketing Specialist 2023-08-20   

   Annual Salary (USD)   5  
0                60000 NaN  
1                48000 NaN

In [ ]:
EmployeeData.groupby('Department')['Annual Salary (USD)'].sum()
#aqui se muestra la suma de salario por departamento

Department
Customer Service            203000
Finance                     233000
Human Resources             200000
Information Technology      253000
Marketing                   203000
Operations                  210000
Production                  224000
Quality Assurance           235000
Research and Development    218000
Sales                       225000
Name: Annual Salary (USD), dtype: int64

| Función              | Qué hace                                   |
|----------------------|---------------------------------------------|
| `.sum()`             | Suma los valores de cada grupo              |
| `.mean()`            | Calcula el promedio                         |
| `.count()`           | Cuenta los registros                        |
| `.min()` / `.max()`  | Mínimo / máximo                             |
| `.agg()`             | Permite aplicar múltiples funciones         |
| `.size()`            | Cuenta el tamaño de cada grupo              |
| `.first()` / `.last()` | Primer / último valor del grupo           |


In [ ]:
EmployeeData.groupby('Department').agg({
    'Annual Salary (USD)': ['mean','max','min','sum']
})


Annual Salary (USD)                      
                                        mean    max    min     sum
Department                                                        
Customer Service                     50750.0  68000  40000  203000
Finance                              58250.0  70000  48000  233000
Human Resources                      50000.0  60000  42000  200000
Information Technology               63250.0  75000  48000  253000
Marketing                            50750.0  65000  42000  203000
Operations                           52500.0  65000  42000  210000
Production                           56000.0  72000  40000  224000
Quality Assurance                    58750.0  75000  42000  235000
Research and Development             54500.0  70000  42000  218000
Sales                                56250.0  72000  48000  225000

### `lambda` y funciones en `.agg`

In [ ]:
EmployeeData.groupby('Department').agg({
    'Annual Salary (USD)' : ['mean',
                             lambda x : x.max()-x.min()]
})

Annual Salary (USD)           
                                        mean <lambda_0>
Department                                             
Customer Service                     50750.0      28000
Finance                              58250.0      22000
Human Resources                      50000.0      18000
Information Technology               63250.0      27000
Marketing                            50750.0      23000
Operations                           52500.0      23000
Production                           56000.0      32000
Quality Assurance                    58750.0      33000
Research and Development             54500.0      28000
Sales                                56250.0      24000

Nombres personalizados de las columnas por tuplas.

In [ ]:
EmployeeData.groupby('Department').agg(
    #name_col = (col,function)
    Salario_promedio = ('Annual Salary (USD)','mean'),
    Rango_salario = ('Annual Salary (USD)',
                     lambda x : x.max()-x.min())
)

Salario_promedio  Rango_salario
Department                                               
Customer Service                   50750.0          28000
Finance                            58250.0          22000
Human Resources                    50000.0          18000
Information Technology             63250.0          27000
Marketing                          50750.0          23000
Operations                         52500.0          23000
Production                         56000.0          32000
Quality Assurance                  58750.0          33000
Research and Development           54500.0          28000
Sales                              56250.0          24000

In [ ]:
def rango(valores):
    return valores.max() - valores.min()

EmployeeData.groupby('Department').agg(
    #name_col = (col,function)
    Salario_promedio = ('Annual Salary (USD)','mean'),
    Rango_salario = ('Annual Salary (USD)',rango)
)

Salario_promedio  Rango_salario
Department                                               
Customer Service                   50750.0          28000
Finance                            58250.0          22000
Human Resources                    50000.0          18000
Information Technology             63250.0          27000
Marketing                          50750.0          23000
Operations                         52500.0          23000
Production                         56000.0          32000
Quality Assurance                  58750.0          33000
Research and Development           54500.0          28000
Sales                              56250.0          24000

# Manejo de datos faltantes
- **`DataFrame.isnull()`**: Devuelve un DataFrame booleano indicando valores nulos.
- **`DataFrame.notnull()`**: Devuelve un DataFrame booleano indicando valores no nulos.
- **`DataFrame.fillna()`**: Rellena valores nulos con un valor especificado.
- **`DataFrame.dropna()`**: Elimina filas o columnas con valores nulos.

## `DataFrame.isnull()`

Devuelve un DataFrame del mismo tamaño, pero con valores booleanos (`True` o `False`) que indican si una celda es nula (`NaN`). Si `NaN` entonces `True`



```python
df.isnull()

```



In [ ]:
import numpy as np

In [ ]:
df = pd.DataFrame({
    'Nombre': [np.nan, 'Luis', np.nan],
    'Edad': [25, np.nan, 30],
    'Ciudad': ['CDMX', 'GDL', 'MTY']
})

In [ ]:
df.isnull()

Nombre   Edad  Ciudad
0    True  False   False
1   False   True   False
2    True  False   False

In [ ]:
df.isnull().sum() #cuenta valores nulos por columna

Nombre    2
Edad      1
Ciudad    0
dtype: int64

In [ ]:
#Filtra las filas con valores nulos por columna especifica

df[df['Nombre'].isnull()]

Nombre  Edad Ciudad
0    NaN  25.0   CDMX
2    NaN  30.0    MTY

In [ ]:
df.isnull().any() #Verifica si hay al menos un nan

Nombre     True
Edad       True
Ciudad    False
dtype: bool

In [ ]:
df.isnull().all() #Verifica si toda la columna es nan

Nombre    False
Edad      False
Ciudad    False
dtype: bool

##`DataFrame.notnull()`

Devuelve un `DataFrame` del mismo tamaño, pero con valores booleanos que indican si la celda NO es nula, es decir, si tiene un dato válido.

`True` → valor presente (no nulo)

`False` → valor ausente (`NaN`)

In [ ]:
df.notnull()

Nombre   Edad  Ciudad
0   False   True    True
1    True  False    True
2   False   True    True

In [ ]:
df.notnull().sum()

Nombre    1
Edad      2
Ciudad    3
dtype: int64

## `DataFrame.fillna()`

Rellena los valores faltantes (`NaN`) en un DataFrame con un valor específico, o usando una técnica de propagación (como copiar valores anteriores o siguientes).

```python
df.fillna(value=None, method=None, axis=None, inplace=False, limit=None)

```

- `value`:	El valor con el que quieres reemplazar los NaN. Puede ser un escalar, diccionario o Series.

- `method`:	Método de propagación:
    - `'ffill'` (forward fill) : Copia el valor anterior no nulo hacia abajo.
    - `'bfill'` (backward fill): Copia el valor siguiente no nulo hacia arriba.

- `axis`:	`0` para filas (por defecto), `1` para columnas.

- `inplace`: 	Si es `True`, modifica el `DataFrame` original.

- `limit`:	Número máximo de valores consecutivos a llenar.

In [ ]:
df

Nombre  Edad Ciudad
0    NaN  25.0   CDMX
1   Luis   NaN    GDL
2    NaN  30.0    MTY

In [ ]:
df.fillna(0) #Remplazar con un valor constantes o en su defecto:

Nombre  Edad Ciudad
0      0  25.0   CDMX
1   Luis   0.0    GDL
2      0  30.0    MTY

| Tipo de dato     | ¿Se rellena con 0? | Resultado                                                                 |
|------------------|--------------------|---------------------------------------------------------------------------|
| `int`, `float`   | ✅ Sí              | `NaN` se reemplaza con `0`                                                |
| `object` (texto) | ✅ Sí              | Reemplaza `NaN` con cadena `'0'` si puede, o lanza error si no puede convertir |
| `bool`           | ✅ Sí              | `NaN` se convierte en `False` (porque `0` en booleano es `False`)         |
| `datetime`       | ❌ No              | Lanza error si no se puede convertir `0` en una fecha                     |


In [ ]:
#Reemplazar por columna:
df.fillna({'Edad': 26, 'Nombre': 'Desconocido'})

Nombre  Edad Ciudad
0  Desconocido  25.0   CDMX
1         Luis  26.0    GDL
2  Desconocido  30.0    MTY

In [ ]:
df.fillna(method='ffill')

<ipython-input-208-5c0beae7dc1e>:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill')


Nombre  Edad Ciudad
0    NaN  25.0   CDMX
1   Luis  25.0    GDL
2   Luis  30.0    MTY

In [ ]:
df.fillna(method='bfill')

<ipython-input-209-b823574c06e2>:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='bfill')


Nombre  Edad Ciudad
0   Luis  25.0   CDMX
1   Luis  30.0    GDL
2    NaN  30.0    MTY

In [ ]:
df.fillna({'Nombre': 'Sin nombre', 'Edad': df['Edad'].mean()})

Nombre  Edad Ciudad
0  Sin nombre  25.0   CDMX
1        Luis  27.5    GDL
2  Sin nombre  30.0    MTY

### Tipos mixtos

Si tienes tipos mixtos, lo mejor es usar un diccionario para asignar valores distintos según la columna, como:

In [ ]:
df = pd.DataFrame({
    'Nombre': ['Ana', np.nan],
    'Edad': [25, np.nan],
    'Activo': [True, np.nan]
})


In [ ]:
df.fillna({
    'Nombre': 'Sin nombre',
    'Edad': 0,
    'Activo': False
})


<ipython-input-215-b966be673de1>:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df.fillna({


Nombre  Edad  Activo
0         Ana  25.0    True
1  Sin nombre   0.0   False

##`DataFrame.dropna()`

Se utiliza para eliminar filas o columnas que contienen valores nulos (`NaN`).



```python
df.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)

```
- `axis`:	`0` = elimina filas (por defecto), `1` = elimina columnas

- `how`:
    - `'any'`: elimina si hay al menos un `NaN`;
    - `'all'` elimina solo si todos los valores son `NaN`

- `thresh`:	Número mínimo de valores no nulos requeridos para mantener la fila/columna

- `subset`:	Lista de columnas a considerar (ignora el resto)

- `inplace`:	Si es `True`, modifica el `DataFrame` original

In [ ]:
df = pd.DataFrame({
    'Nombre': ['Ana', 'Luis',np.nan],
    'Edad': [25, np.nan, np.nan],
    'Ciudad': ['Lima', 'Bogotá', 'Quito'],
    'Activo':[True,np.nan,np.nan]
})
df

Nombre  Edad  Ciudad Activo
0    Ana  25.0    Lima   True
1   Luis   NaN  Bogotá    NaN
2    NaN   NaN   Quito    NaN

In [ ]:
df.dropna() #Elimina todas las filas con al menos un Nan

Nombre  Edad Ciudad Activo
0    Ana  25.0   Lima   True

In [ ]:
df.dropna(axis=1, how='all') #Eliminar columna si todos son NaN

Nombre  Edad  Ciudad Activo
0    Ana  25.0    Lima   True
1   Luis   NaN  Bogotá    NaN
2    NaN   NaN   Quito    NaN

In [ ]:
df.dropna(subset=['Edad'])  # Elimina filas de Edad que sean NaN

Nombre  Edad Ciudad Activo
0    Ana  25.0   Lima   True

In [ ]:
df.dropna(subset=['Edad','Activo'])#Elimina filas de Edad y Activo que sean nan

Nombre  Edad Ciudad Activo
0    Ana  25.0   Lima   True

In [ ]:
df.dropna(thresh=2)
#mantén solo las filas que tengan al menos 2 valores no nulos

Nombre  Edad  Ciudad Activo
0    Ana  25.0    Lima   True
1   Luis   NaN  Bogotá    NaN

#Operaciones estadísticas y matemáticas
- **`DataFrame.describe()`**: Proporciona estadísticas descriptivas del DataFrame.
- **`DataFrame.mean()`**: Calcula la media de las columnas o filas.
- **`DataFrame.median()`**: Calcula la mediana de las columnas o filas.
- **`DataFrame.sum()`**: Calcula la suma de las columnas o filas.
- **`DataFrame.cumsum()`**: Calcula la suma acumulativa de las columnas o filas.
- **`DataFrame.min()`**: Encuentra el valor mínimo de las columnas o filas.
- **`DataFrame.max()`**: Encuentra el valor máximo de las columnas o filas.
- **`DataFrame.corr()`**: Calcula la correlación entre columnas.
- **`DataFrame.cov()`**: Calcular la matriz de covarianza de un DataFrame.
- **`DataFrame.count()`**: Cuenta el número de valores no nulos en cada columna o fila.

##`DataFrame.describe()`
Genera estadísticas descriptivas de las columnas del DataFrame.



```python
#sintaxis
df.describe(include='number',exclude=None,percentiles=[0.25,0.5,0.75])
```



In [ ]:
EmployeeData.describe()

Hire Date  Annual Salary (USD)             5
count                   40            40.000000      2.000000
mean   2023-07-10 13:12:00         55100.000000  24025.000000
min    2023-01-17 00:00:00         40000.000000     50.000000
25%    2023-04-25 12:00:00         48000.000000  12037.500000
50%    2023-07-11 00:00:00         53500.000000  24025.000000
75%    2023-09-22 12:00:00         62750.000000  36012.500000
max    2023-12-03 00:00:00         75000.000000  48000.000000
std                    NaN         10683.800342  33905.770158

##`DataFrame.mean()`



```python
df.mean(axis=0, skipna=True, numeric_only=False)

```
- `axis=0`:	Calcula el promedio por columnas (default). Usa `axis=1` para hacerlo por filas.

- `skipna=True`:	Ignora los valores nulos `NaN`. Si lo pones en `False`, la media será `NaN` si hay datos faltantes.

- `numeric_only=False`	Si está en `True`, calcula la media solo de columnas numéricas. Si hay objetos o fechas, las ignora automáticamente.


In [ ]:
EmployeeData.mean(numeric_only=True)

Annual Salary (USD)    55100.0
5                      24025.0
dtype: float64

## `DataFrame.median()`

Calcula la mediana (el valor central de una distribución ordenada).

```python
DataFrame.median(axis=0, skipna=True, numeric_only=False)
```




In [ ]:
EmployeeData.median(numeric_only=True)

Annual Salary (USD)    53500.0
5                      24025.0
dtype: float64

## `DataFrame.sum()`



```python
DataFrame.sum(axis=0, skipna=True, numeric_only=False)
```

In [ ]:
EmployeeData.sum(numeric_only=True)

Annual Salary (USD)    2204000.0
5                        48050.0
dtype: float64

## `DataFrame.cumsum()`

Devuelve la suma acumulativa de los valores a lo largo de un eje. Es decir, va sumando fila por fila o columna por columna, acumulando el total progresivamente.

```python
DataFrame.cumsum(axis=0, skipna=True, numeric_only=False)
```


In [ ]:
df = pd.DataFrame({
    'Ventas_A': [100, 200, 300],
    'Ventas_B': [50, 150, 100]
})

In [ ]:
df.cumsum()

Ventas_A  Ventas_B
0       100        50
1       300       200
2       600       300

In [ ]:
df.cumsum(axis=1)

Ventas_A  Ventas_B
0       100       150
1       200       350
2       300       400

In [ ]:
df2 = pd.DataFrame({
    'A': [1, None, 3],
    'B': [4, 5, None],
    'C': [4,10,None],
    'D':[None,3,4]
})
df2

A    B     C    D
0  1.0  4.0   4.0  NaN
1  NaN  5.0  10.0  3.0
2  3.0  NaN   NaN  4.0

In [ ]:
df2.cumsum()

A    B     C    D
0  1.0  4.0   4.0  NaN
1  NaN  9.0  14.0  3.0
2  4.0  NaN   NaN  7.0

Con `skipna=False`, cualquier `NaN` detiene la acumulación:

In [ ]:
df2.cumsum(axis=1, skipna=False)

A    B    C   D
0  1.0  5.0  9.0 NaN
1  NaN  NaN  NaN NaN
2  3.0  NaN  NaN NaN

##`DataFrame.min()` `DataFrame.max()`

```python
DataFrame.min(axis=0, skipna=True, numeric_only=False)

```

In [ ]:
df = pd.DataFrame({
    'A': [3, 7, 2],
    'B': [5, 1, 9],
    'C': [8, 4, 6]
})
df

A  B  C
0  3  5  8
1  7  1  4
2  2  9  6

In [ ]:
df.min()

A    2
B    1
C    4
dtype: int64

In [ ]:
df.max()

A    7
B    9
C    8
dtype: int64

In [ ]:
df.min(axis=1)

0    3
1    1
2    2
dtype: int64

In [ ]:
df.max(axis=1)

0    8
1    7
2    9
dtype: int64

## `DataFrame.corr()`



```python
DataFrame.corr(method='pearson', numeric_only=_NoDefault.no_default, min_periods=1)

```
- `method:`
    - `'pearson'`: (por defecto): relación lineal.
    - `'kendall'` : ordenada, no lineal.
    - `'spearman'`: basada en rangos, útil para relaciones monótonas.

- `numeric_only`: si `True`, calcula solo con columnas numéricas.
- `min_periods`: número mínimo de observaciones requeridas entre pares de columnas para calcular la correlación.


In [ ]:
df = pd.DataFrame({
    'A': [1, 2, 3, 4],
    'B': [4, 3, 2, 1],
    'C': [1, 2, 3, 4]
})

df

A  B  C
0  1  4  1
1  2  3  2
2  3  2  3
3  4  1  4

In [ ]:
df.corr()

A    B    C
A  1.0 -1.0  1.0
B -1.0  1.0 -1.0
C  1.0 -1.0  1.0

In [ ]:
df = pd.DataFrame({
    'A': [1, 2, None, 4, 5],
    'B': [4, None, 2, 1, 3],
    'C': [1, 2, 3, 4, 5]
})
df

A    B  C
0  1.0  4.0  1
1  2.0  NaN  2
2  NaN  2.0  3
3  4.0  1.0  4
4  5.0  3.0  5

##`DataFrame.cov()`



```python
DataFrame.cov(min_periods=None)

```



In [ ]:
df = pd.DataFrame({
    'A': [1, 2, 3, 4, 5],
    'B': [5, 4, 3, 2, 1],
    'C': [2, 3, 4, 5, 6]
})
df

A  B  C
0  1  5  2
1  2  4  3
2  3  3  4
3  4  2  5
4  5  1  6

In [ ]:
df.cov()

A    B    C
A  2.5 -2.5  2.5
B -2.5  2.5 -2.5
C  2.5 -2.5  2.5

## `DataFrame.count()`

Cuenta la cantidad de valores no nulos (non-NaN) en cada columna (por defecto) o fila de un DataFrame.



```python
DataFrame.count(axis=0, level=None, numeric_only=False)
```

- `level` (opcional): si el DataFrame tiene un índice jerárquico (MultiIndex), puedes especificar un nivel para hacer la cuenta a lo largo de ese nivel.

In [ ]:
df = pd.DataFrame({
    'A': [1, 2, np.nan, 4],
    'B': [np.nan, 2, 3, 4],
    'C': ['a', 'b', 'c', None]
})

df

A    B     C
0  1.0  NaN     a
1  2.0  2.0     b
2  NaN  3.0     c
3  4.0  4.0  None

In [ ]:
df.count()

A    3
B    3
C    3
dtype: int64

In [ ]:
df.notnull().sum()

A    3
B    3
C    3
dtype: int64

#Funciones de aplicación
- **`DataFrame.apply()`**: Aplica una función a lo largo de un eje del DataFrame (filas o columnas).
- **`DataFrame.applymap()`**: Aplica una función a cada elemento del DataFrame.
- **`Series.map()`**: Aplica una función a cada elemento de una Serie.

## `DataFrame.apply()`

Sirve para aplicar una función a lo largo de un eje (filas o columnas) de un DataFrame.



```python
#sintaxis
DataFrame.apply(func, axis=0, raw=False, result_type=None, args=(), **kwargs)

```
- `func`: la función que quieres aplicar. Puede ser:
    - Una función de Python (`sum`, `len`, `np.mean`, etc.)
    - Una función lambda
    - Una función personalizada que tú definas.

- `axis`:
    - `0` (o `'index'`): aplica la función a lo largo de las columnas (es decir, la función recibe cada columna como una Series)
    - `1` (o `'columns'`): aplica la función a lo largo de las filas

- `raw:`
    - Si es `True`, pasa los valores como un arreglo ndarray sin etiquetas.
    - Si es `False` (default), pasa cada columna o fila como una Series con etiquetas.
- `args=()`: Permite pasar argumentos adicionales a la función que estás aplicando. Debes colocarlos como una tupla.

- `**kwargs`: Te permite pasar argumentos con nombre (keywords) adicionales a tu función.


In [3]:
df = pd.DataFrame({
    'A': [1, 2, 3],
    'B': [10, 20, 30]
})
df

A   B
0  1  10
1  2  20
2  3  30

In [4]:
df.apply(sum) #por columnas

A     6
B    60
dtype: int64

In [7]:
#por filas usando una funcion lambda
df.apply(lambda rows: rows['A']+rows['B'], axis=1)

#similar
# df.apply(sum,axis=1)

0    11
1    22
2    33
dtype: int64

In [14]:
def rango(series):
  return series.max()-series.min()

df.apply(rango)

A     2
B    20
dtype: int64

In [16]:
df = pd.DataFrame({'A': [1, 2], 'B': [3, 4]})
df

A  B
0  1  3
1  2  4

In [18]:
df.apply(lambda row: [row['A']+1, row['B']+1], axis=1, result_type='expand')

0  1
0  2  4
1  3  5

In [22]:
def multiplicar(x, factor):
    return x * factor

df = pd.DataFrame({'A': [1, 2, 3]})

df['A'].apply(multiplicar, args=(10,))
# Esto es equivalente a que Pandas llame: multiplicar(valor_de_A, 10)


0    10
1    20
2    30
Name: A, dtype: int64

In [20]:
def multiplicador(x, factor, sumando):
    return x * factor + sumando

df.apply(multiplicador,args=(10,1))

A
0  11
1  21
2  31

`**kwargs`

In [23]:
def sumar_valor(x, incremento=0):
    return x + incremento

df['A'].apply(sumar_valor, incremento=5)

0    6
1    7
2    8
Name: A, dtype: int64

##`DataFrame.applymap()`

Para aplicar una función elemento por elemento, es decir, a cada valor individual del DataFrame.



```python
#sintaxis
DataFrame.applymap(func)
```


In [25]:
df = pd.DataFrame({
    'A': [1, -2, 3],
    'B': [-4, 5, -6]
})

# Aplicar valor absoluto a cada elemento
df.applymap(abs)

<ipython-input-25-8398ab8ad34a>:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df.applymap(abs)


A  B
0  1  4
1  2  5
2  3  6

In [28]:
from math import sqrt

df.applymap(lambda x: sqrt(x**2))

<ipython-input-28-61922dcf5072>:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df.applymap(lambda x: sqrt(x**2))


A    B
0  1.0  4.0
1  2.0  5.0
2  3.0  6.0

##`Series.map()`
Aplicar una funcion a cada elemento de una serie.



```python
#sintaxis
Series.map(arg, na_action=None)
```
- `arg` puede ser:
    - una función → para transformar cada valor
    - un diccionario → para reemplazar valores
    - otra Series → para mapear con base en valores clave

- `na_action='ignore'` → evita aplicar la función a valores `NaN`




In [29]:
s = pd.Series([1, 2, 3])
s.map(lambda x: x ** 2)

0    1
1    4
2    9
dtype: int64

In [30]:
s = pd.Series(['cat', 'dog', 'rabbit'])
s.map({'cat': 'gato', 'dog': 'perro'}) #Nan a valor faltante

0     gato
1    perro
2      NaN
dtype: object

In [31]:
s.replace({'cat': 'gato', 'dog': 'perro'}) #incluye el valor original

0      gato
1     perro
2    rabbit
dtype: object

In [32]:
s1 = pd.Series(['a', 'b', 'c'])
s2 = pd.Series({'a': 1, 'b': 2, 'c': 3})
s1.map(s2)

0    1
1    2
2    3
dtype: int64

# Funciones de ventana y de series temporales
- **`DataFrame.rolling()`**: Proporciona operaciones de ventana móvil.
- **`DataFrame.expanding()`**: Proporciona operaciones acumulativas.
- **`DataFrame.resample()`**: Remuestrea series temporales a una frecuencia diferente.
- **`pd.date_range()`**: Genera un rango de fechas.
- **`pd.to_datetime()`**: Convierte una cadena u objeto a tipo datetime.

## `DataFrame.rolling()`

Aplica una función agregada (como la media, la suma, etc.) sobre una ventana de elementos consecutivos en una serie de datos.



```python
#sintaxis
DataFrame.rolling(window, min_periods=None, axis=0, method='single')

```
- `window`: El tamaño de la ventana, que indica el número de observaciones a considerar en cada paso de la ventana deslizante.
    - Ejemplo: `window=4` significa que la operación se realizará sobre 4 filas consecutivas de datos.

- `min_periods` (opcional):  El número mínimo de observaciones que deben estar presentes en la ventana para realizar la operación. Si es menor que el tamaño de la ventana, se llenan los valores faltantes (`NaN`) al inicio de la serie.
    - `min_periods=2` significa que, para aplicar una función como la media, la ventana debe tener al menos 2 valores no nulos.
    
- `method` (opcional):  Define el tipo de ventana a usar. El valor predeterminado es 'single', pero puede ser 'table' si estás aplicando la ventana sobre tablas de datos multidimensionales (poco común).

In [33]:
# Datos de ejemplo: ventas trimestrales de una empresa
data = {'Año': [2021, 2021, 2021, 2021, 2022, 2022, 2022, 2022],
        'Trimestre': ['Q1', 'Q2', 'Q3', 'Q4', 'Q1', 'Q2', 'Q3', 'Q4'],
        'Ventas': [100, 120, 130, 140, 150, 160, 170, 180]}

df = pd.DataFrame(data)
df

Año Trimestre  Ventas
0  2021        Q1     100
1  2021        Q2     120
2  2021        Q3     130
3  2021        Q4     140
4  2022        Q1     150
5  2022        Q2     160
6  2022        Q3     170
7  2022        Q4     180

In [44]:
# Cálculo de la media móvil de 4 trimestres
df['Promedio_Móvil'] = df['Ventas'].rolling(window=4).mean()

print(df)

    Año Trimestre  Ventas  Promedio_Móvil
0  2021        Q1     100             NaN
1  2021        Q2     120             NaN
2  2021        Q3     130             NaN
3  2021        Q4     140           122.5
4  2022        Q1     150           135.0
5  2022        Q2     160           145.0
6  2022        Q3     170           155.0
7  2022        Q4     180           165.0


`ventana.funcion` --> `rolling.function`
- La ventana de tamaño 4 (cuatro trimestres) se mueve sobre las columnas `Ventas` para calcular el promedio de cada conjunto de 4 trimestres consecutivos.

- En las primeras tres filas, el promedio móvil no puede calcularse porque no hay suficientes datos anteriores (menos de 4 trimestres).

- A partir del cuarto trimestre, puedes ver el promedio de las ventas de los últimos 4 trimestres.

##`DataFrame.expanding()`

Se usa para aplicar funciones acumulativas sobre los datos desde el inicio hasta la fila actual, es decir, una ventana creciente. A diferencia de `.rolling()`, que usa una ventana fija, `.expanding()` comienza con un número mínimo de observaciones y se expande hacia abajo a medida que avanzamos en el `DataFrame`.



```python
#sintaxis
DataFrame.expanding(min_periods=1, axis=0, method='single')
```

- `min_periods` (por defecto = 1): Número mínimo de observaciones requeridas para que se calcule el resultado. Si no se alcanzan, el resultado será `NaN`

- `axis` (por defecto = 0): Eje sobre el que se aplicará la operación.
    - `0 `→ vertical (por filas),
    - `1` → horizontal (por columnas)

- `method` (por defecto = 'single'):


In [45]:
# Ventas mensuales simuladas
data = {'Mes': ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo'],
        'Ventas': [100, 120, 150, 130, 160]}

df = pd.DataFrame(data)

df

Mes  Ventas
0    Enero     100
1  Febrero     120
2    Marzo     150
3    Abril     130
4     Mayo     160

In [50]:
# Promedio acumulado de ventas
df['Promedio_Acumulado'] = df['Ventas'].expanding(min_periods=1).mean()
df

Mes  Ventas  Promedio_Acumulado
0    Enero     100          100.000000
1  Febrero     120          110.000000
2    Marzo     150          123.333333
3    Abril     130          125.000000
4     Mayo     160          132.000000

- El promedio acumulado se calcula incluyendo todas las ventas desde el inicio hasta el mes actual.

- En la segunda fila: (100 + 120) / 2 = 110.0

- En la tercera fila: (100 + 120 + 150) / 3 = 123.33...

- Así se va expandiendo la ventana hasta el final.

##`DataFrame.resample()`

Permite agrupar datos temporales por frecuencia (diaria, mensual, anual, etc.) y aplicar funciones de agregación como `mean()`, `sum()`, `count()`, etc.

```python
DataFrame.resample(rule,
                   axis=0,
                   closed=None,
                   label=None,
                   convention='start',
                   kind=None,
                   loffset=None, base=None, on=None, level=None)

```
- `rule` (obligatorio): Frecuencia para reagrupar los datos.
    - `'D'`: diario
    - `'W'`: semanal
    - `'M'`: mensual
    - `'Q'`: trimestral
    - `'A'`: anual
    - `'H'`: por hora

- `on` (opcional): Si el índice no es un datetime, puedes usar este parámetro para indicar qué columna contiene las fechas.


In [51]:
# Crear un DataFrame de visitas diarias
fechas = pd.date_range(start="2023-01-01", periods=10, freq='D')
visitas = [120, 135, 150, 160, 180, 145, 155, 170, 190, 210]

df = pd.DataFrame({'Fecha': fechas, 'Visitas': visitas})
df

Fecha  Visitas
0 2023-01-01      120
1 2023-01-02      135
2 2023-01-03      150
3 2023-01-04      160
4 2023-01-05      180
5 2023-01-06      145
6 2023-01-07      155
7 2023-01-08      170
8 2023-01-09      190
9 2023-01-10      210

In [52]:
# Reagrupar por semana y sumar visitas
df_semanal = df.resample('W', on='Fecha').sum()

print(df_semanal)

            Visitas
Fecha              
2023-01-01      120
2023-01-08     1095
2023-01-15      400


## `pd.date_range()`
Crea una secuencia de fechas con una frecuencia y duración determinadas.



```python
pd.date_range(start=None, end=None, periods=None, freq='D', tz=None, normalize=False, name=None, inclusive=None)

```
- `start` (`str` o `datetime`): Fecha de inicio del rango.

- `end` (`str` o `datetime`): Fecha de fin del rango.

- `periods` (`int`): Número de fechas a generar. Si lo defines, no necesitas `end`.

- `freq` (`str`): Frecuencia entre fechas.
    - `'D'`: dias
    - `'W'`: semanal
    - `'M'`: fin de mes
    - `'Q'`: fin de trimestre
    - `'Y'`: fin de año
    - `'H'`: por hora
    - `'B'`: días hábiles

- `tz` (`str`):  Zona horaria. Ej: 'UTC', 'America/New_York'.

- `normalize` (`bool`): Si es `True`, ajusta todas las fechas al inicio del día (00:00:00)



In [57]:
fechas = pd.date_range(start='2023-01-01', end='2023-01-10')
pd.Series(fechas,name='Time')

0   2023-01-01
1   2023-01-02
2   2023-01-03
3   2023-01-04
4   2023-01-05
5   2023-01-06
6   2023-01-07
7   2023-01-08
8   2023-01-09
9   2023-01-10
Name: Time, dtype: datetime64[ns]